fix(subscriber): record timestamps for updates last

hawkw opened the pull request:
## Motivation

Currently, when constructing an update message to send over the wire, a
timestamp is taken first, and then the protobuf data is constructed.
This can lead to issues where the "now" timestamp is actually _before_
timestamps present in the stats sent in that update, since the stats for
a particular task/resource/async op might be updated on another thread
after taking the update's "now" timestamp. This results in issues like
#266.

## Solution

There's no actual reason to take those timestamps *before* we assemble
the update. This branch changes the aggregator to build all the various
data updates in an update message, and *then* record the update's "now"
timestamp. Any timestamps for tasks/resources/async ops that are
recorded after the update's "now" timestamp will now be included in the
*next* update.

Fixes #266
Depends on #290

hawkw committed:
refac(subscriber): factor out update construction

This branch refactors how the aggregator task builds update messages. In
particular, it factors out the shared code between building the initial
update for a subscription and building the subsequent updates in
`publish` into methods that are called with an `Include` enum to
determine whether the update should include all tasks/resources/ops, or
just the ones since the last update. This makes the code in
`add_instrument_subscription` and in `publish` much easier to read.

This is just the refactor part from #289 split out into a separate
branch; this should make no functional change. If we decide that #289 is
the correct way to fix the bug, we can rebase that branch onto this, so
that it _just_ includes the bugfix. This way, we can land the refactor
anyway, because I think it's nice.

seanmonstar commented:
> There's no actual reason to take those timestamps before we assemble the update.

I'd think the reason for taking the timestamp first is because all the other stuff can take some time to compute, and we want to record as close as possible when the event actually happened, not when we finished understanding what the event means.

hawkw commented:
> > There's no actual reason to take those timestamps before we assemble the update.
> 
> I'd think the reason for taking the timestamp first is because all the other stuff can take some time to compute, and we want to record as close as possible when the event actually happened, not when we finished understanding what the event means.

Right, but the issue is that converting the data to the protobuf format takes time, and the data may be updated while we're doing that. If the data is updated by another thread after when we take the "now" timestamp and when we actually construct the protobuf representation of that data, it might have a timestamp that occurs _after_ "now". #266 explains why this is a problem.

We could, alternatively, handle this kind of thing in the console UI; this _felt_ like a better solution to me, though, since it wouldn't have to be solved again in another consumer of the wire format.

hawkw committed:
fix(subscriber): record timestamps for updates last

## Motivation

Currently, when constructing an update message to send over the wire, a
timestamp is taken first, and then the protobuf data is constructed.
This can lead to issues where the "now" timestamp is actually _before_
timestamps present in the stats sent in that update, since the stats for
a particular task/resource/async op might be updated on another thread
after taking the update's "now" timestamp. This results in issues like
#266.

## Solution

There's no actual reason to take those timestamps *before* we assemble
the update. This branch changes the aggregator to build all the various
data updates in an update message, and *then* record the update's "now"
timestamp. Any timestamps for tasks/resources/async ops that are
recorded after the update's "now" timestamp will now be included in the
*next* update.

Fixes #266
Depends on #288

hawkw commented:
I went ahead and split out the refactoring into a separate branch (#290) so we can go ahead and merge that regardless of what we decide is the best way to fix the bug.

hawkw committed:
Merge branch 'main' into eliza/late-nows

thombles commented:
I've been fiddling with this crash in parallel (I can trigger it pretty reliably) and was investigating the same change. I tried checking out the tip of this branch (713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6) and I'm still getting a similar crash/backtrace sadly.

[console-crash-713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6.txt](https://github.com/tokio-rs/console/files/8092430/console-crash-713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6.txt)

hawkw commented:
> I've been fiddling with this crash in parallel (I can trigger it pretty reliably) and was investigating the same change. I tried checking out the tip of this branch ([713a7be](https://github.com/tokio-rs/console/commit/713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6)) and I'm still getting a similar crash/backtrace sadly.
> 
> [console-crash-713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6.txt](https://github.com/tokio-rs/console/files/8092430/console-crash-713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6.txt)

Hmm, that's a shame that this didn't fix it. Can you share how you're able to reliably trigger the panic?

thombles commented:
Not really, sorry. My app under test just has 50–70 long-running tasks ticking along mostly idle and it crashes within 10 minutes or so looking at at the tasks list view.

thombles commented:
If it helps account for any particular skew, the instrumented device is an iPhone I'm connected to over WiFi, with console running on my Mac.

hawkw commented:
@thombles just to confirm, the iPhone's `console-subscriber` dependency is from this branch on commit 713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6? Not just the version of the console UI running on your mac?

hawkw commented:
Oh, one other question: when the panic occurs, do you know if it's _always_ in `AsyncOpsState::retain_active`, or does it ever panic in `TasksState::retain_active` or `ResourcesState::retain_active`?

thombles commented:
> @thombles just to confirm, the iPhone's `console-subscriber` dependency is from this branch on commit [713a7be](https://github.com/tokio-rs/console/commit/713a7befc3dcb9b7dae5a6e5b4fbe220e47a6eb6)? Not just the version of the console UI running on your mac?

Whoops, great guess - yes I forgot to update the subscriber end and obviously this code is in the subscriber crate. Sorry to mislead. I've now updated both device to `main` and it's rock solid. It does make me wonder whether console should defensively do a saturating duration_since just in case it gets bogus data but I guess it's not critical?

> when the panic occurs, do you know if it's _always_ in `AsyncOpsState::retain_active`, or does it ever panic in `TasksState::retain_active` or `ResourcesState::retain_active`?

So far I've only noticed it in the `AsyncOpsState` but I haven't analysed whether there's any special reason for that.

hawkw commented:
@thombles 
> Whoops, great guess - yes I forgot to update the subscriber end and obviously this code is in the subscriber crate. Sorry to mislead. I've now updated both device to `main` and it's rock solid.

Ah, yup, there's the problem --- thanks for double-checking. So, this change probably does fix the issue, which is good, because I felt like I was losing my mind there for a second. :)

> It does make me wonder whether console should defensively do a saturating duration_since just in case it gets bogus data but I guess it's not critical?

Yeah, it probably should — we should give the whole console UI a pass for potential panics in the face of weird data from the instrumented application.
